Load file from credentials file (~/.aws/credentials)

daddykotex opened the pull request:
While working on #722 I realized we could not pull the info from `~/.aws/credentials` so I added it.

what do you think?

daddykotex committed:
Add a way to load credentials from config file

daddykotex committed:
Load file from credentials file

daddykotex committed:
Sort imports

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsProvider.scala:
```diff
@@ -36,24 +39,33 @@ object AwsCredentialsProvider {
+    val env = Resource.eval(fromEnv[F]).map(F.pure)
+    val fromDisk = Resource.eval(loadFromDisk[F]).map(F.pure)
+    env
+      .orElse(fromDisk)
```
`fromDisk` should be lowest priority. On local setups, solutions like [aws-vault](https://github.com/99designs/aws-vault) or [finto](https://github.com/threadwaste/finto) exist to emulate the AWS runtimes and should take precedence over the config filec.

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+    Resource
+      .fromAutoCloseable(
+        Sync[F].delay(
+          scala.io.Source
```
Unfortunately, this is not gonna be working on node environments. Let's use [fs2.io](https://fs2.io/#/io?id=files) instead, that'll work on the three environments.

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+      .fromAutoCloseable(
+        Sync[F].delay(
+          scala.io.Source
+            .fromFile(Paths.get(home, ".aws", "credentials").toUri())
```
let's not hardcode this here, but inject the `fs2.io.files.Path` indicating the config file all the way here

Baccata commented on modules/aws-http4s/src/smithy4s/aws/http4s/package.scala:
```diff
@@ -27,7 +27,7 @@ package object http4s {
   ) {
 
-    def awsClient[F[_]: Temporal](
+    def awsClient[F[_]: Async](
```
this is not binary compatible but we're not making a strong promise regarding the AWS modules at this point, moreover I'm pretty confident it's unlikely to break any user code .

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+  }
+}
+
+object Profile {
```
this object and its contents should be private

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+)
+
+object AwsCredentialsFile {
+  def loadFromDisk[F[_]](home: String, profile: Option[String])(implicit
```
don't put `home` there, but rather the full location of the file (using fs2.io.files.Path).

Baccata replied:
Also, let's add a test for this with a temp file (created with fs2.io) to ensure that it works on the 3 platforms

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+            creds.profiles
+              .get(p)
+              .toRight(
+                new RuntimeException(
```
let's create actual exceptions for this

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+      )
+  }
+
+  private[aws] def processFileLines(lines: List[String]): AwsCredentialsFile = {
```
I'm pretty certain that this can fail, the type should reflect that

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+      F: Sync[F]
+  ): F[AwsCredentials] = {
+    loadFile(home)
+      .use(lines => processFileLines(lines).pure[F])
```
It's unlikely you can safely `pure` it. It should be `F.delay()` to have a chance to catch the exceptions. Alternatively, the `processFileLines` should return `F` or `Either`

kubukoz replied:
`use` will catch it, but yeah it's better to  `delay` for extra sanity

daddykotex replied:
wrapped it all up in `Either`s

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+          .getOrElse(
+            creds.default
+              .toRight(
+                new RuntimeException(
```
ditto https://github.com/disneystreaming/smithy4s/pull/756/files#r1084974649

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsFile.scala:
```diff
@@ -0,0 +1,158 @@
+        profile
+          .map(p =>
+            creds.profiles
+              .get(p)
```
you should be able to call `getOrElse(default)` there. In any case, it's too much [hadouken](https://tse3.mm.bing.net/th?id=OIP.CEQDgPo1GeSaTXpZXudjtwHaFQ&pid=Api) for my taste

daddykotex replied:
IMO, if you ask for profile X and it's not found, it should fail

not fallback to the default without you knowing

daddykotex replied:
just let me know what you prefer and I'll get to it

but as an example, the AWS sdk does the following if you have a `aws_profile` env var, but not a corresponding profile in the credentials file:
```
ProfileCredentialsProvider(profileName=MY_PROFILE_SSO, profileFile=ProfileFile(profilesAndSectionsMap=[{default=Profile(name=default, properties=[region])}, {}])): Profile file contained no credentials for profile 'MY_PROFILE_SSO': ProfileFile(profilesAndSectionsMap=[{default=Profile(name=default, properties=[region])}, {}])
```

Baccata replied:
> IMO, if you ask for profile X and it's not found, it should fail

Agreed.

Baccata commented on modules/aws/src/smithy4s/aws/AwsCredentialsProvider.scala:
```diff
@@ -36,24 +39,33 @@ object AwsCredentialsProvider {
+    either.liftTo[F]
+  }
+
+  def loadFromDisk[F[_]](implicit F: Sync[F]): F[AwsCredentials] = {
```
let's just call this `fromDisk` (drop the `load`, it's implied), taking a `Profile` for the profile as a parameter. Then create a `getProfileFromEnv[F[_]]` method (profiles may be fetched from elsewhere, a config file or whatnot).  Then wire these methods in the `default` method. 


Also, let's make a newtype for `Profile` (look at `AwsRegion`) for inspiration.

daddykotex replied:
I encoded it as `def getProfileFromEnv: Option[String]` because `F` is not required. Unless we change `SysEnv.envValue` to return a `F`

kubukoz replied:
CE went with `F` for env reading, YMMV though

Baccata commented:
The initiative is very welcome, however it needs polishing, in particular for cross-platform usage. 

Generally speaking: and in particular in native/sjs, when it comes to using the java stdlib (or the scala stdlib that delegates to the java one for I/O stuff), it's likely that things will compile, but it may happen that they don't link. 

In this case, scala-native does implement a subset of java.nio (using pure scala that complies with the same signatures as the java implementation) for files, but it's not the case for sjs.

daddykotex commented:
> The initiative is very welcome, however it needs polishing, in particular for cross-platform usage.
> 
> Generally speaking: and in particular in native/sjs, when it comes to using the java stdlib (or the scala stdlib that delegates to the java one for I/O stuff), it's likely that things will compile, but it may happen that they don't link.
> 
> In this case, scala-native does implement a subset of java.nio (using pure scala that complies with the same signatures as the java implementation) for files, but it's not the case for sjs.

100%, I opened the PR as soon as I got something working to avoid spending too much time on the implementation in case we decided to drop it. Good to know, it could make it's way in!

I'll make the changes that you highlighted, thanks for the thorough review.

At first, I implemented it with fs2-io but I had to add a dependency and so I decided to roll it back. But I did not think that it would buy me the multi-platform ticket!

Baccata commented:
> add a dependency and so I decided to roll it back 

yeah the line is very gray there, but I think the reality is that I'm actually contemplating merging `aws` and `aws-http4s` in a single module because we're gonna need `fs2.Stream` in `aws` and it's silly to not acknowledge that http4s is likely gonna be the only backend we ever support for that particular stuff.

daddykotex committed:
Merge remote-tracking branch 'origin/series/0.17' into dfrancoeur/read-from-file

daddykotex committed:
Take PR reviews comment into account

1. use fs2-io to read the file for multiplatform support
2. wrap throwing code in Either
3. use custom exception to signal error when loading credentials
4. add a test that read/write a file

daddykotex committed:
Remove dummy error

kubukoz commented:
> yeah the line is very gray there, but I think the reality is that I'm actually contemplating merging aws and aws-http4s in a single module because we're gonna need fs2.Stream in aws and it's silly to not acknowledge that http4s is likely gonna be the only backend we ever support for that particular stuff.

I'm of two minds on this, on one hand that's true, but having a split allows users to reuse some of our work in case they want to support another backend. I guess we can always split it again, if someone asks for that.

Baccata commented:
> but having a split allows users to reuse some of our work in case they want to support another backend 

I was thinking that at some point, but now I don't really want to optimise for those hypothetical users. If they disagree with http4s they'll disagree with fs2 most likely, and abstracting over streaming libraries is not a thought I want to entertain.
